vesting/src/lib.rs

```diff
@@ -0,0 +1,261 @@
+		VestingSchedule = VestingScheduleOf<T>
+	{
+		/// Added new vesting schedule (from, to, vesting_schedule)
+		VestingScheduleAdded(AccountId, AccountId, VestingSchedule),
+		/// Claimed vesting (who, locked_amount)
+		Claimed(AccountId, Balance),
+		/// Updated vesting schedules (who)
+		VestingSchedulesUpdated(AccountId),
+	}
+);
+
+decl_error! {
+	/// Error for vesting module.
+	pub enum Error for Module<T: Trait> {
+		ZeroVestingPeriod,
+		ZeroVestingPeriodCount,
+		NumOverflow,
+		InsufficientBalanceToLock,
+		HasNonVestingLocks,
+	}
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		pub fn claim(origin) {
+			let who = ensure_signed(origin)?;
+			let locked_amount = Self::do_claim(&who);
+
+			Self::deposit_event(RawEvent::Claimed(who, locked_amount));
+		}
+
+		pub fn add_vesting_schedule(
+			origin,
+			dest: <T::Lookup as StaticLookup>::Source,
+			schedule: VestingScheduleOf<T>,
+		) {
+			let from = ensure_signed(origin)?;
+			let to = T::Lookup::lookup(dest)?;
+			Self::do_add_vesting_schedule(&from, &to, schedule.clone())?;
+
+			Self::deposit_event(RawEvent::VestingScheduleAdded(from, to, schedule));
+		}
+
+		pub fn update_vesting_schedules(
+			origin,
+			who: <T::Lookup as StaticLookup>::Source,
+			vesting_schedules: Vec<VestingScheduleOf<T>>
+		) {
+			ensure_root(origin)?;
+
+			let account = T::Lookup::lookup(who)?;
+			Self::do_update_vesting_schedules(&account, vesting_schedules)?;
+
+			Self::deposit_event(RawEvent::VestingSchedulesUpdated(account));
+		}
+	}
+}
+
+const VESTING_LOCK_ID: LockIdentifier = *b"vestingm";
+
+impl<T: Trait> Module<T> {
+	fn do_claim(who: &T::AccountId) -> BalanceOf<T> {
+		if let Some((amount, until)) = Self::locked(who) {
+			T::Currency::set_lock(VESTING_LOCK_ID, who, amount, until, WithdrawReasons::all());
+			amount
+		} else {
+			T::Currency::remove_lock(VESTING_LOCK_ID, who);
+			Zero::zero()
+		}
+	}
+
+	/// Returns locked balance info based on current block number: if no remaining locked balance,
+	/// returns `None`, or returns `Some((amount, until))`
+	fn locked(who: &T::AccountId) -> Option<(BalanceOf<T>, T::BlockNumber)> {
+		let now = <frame_system::Module<T>>::block_number();
+		Self::vesting_schedules(who).iter().fold(None, |acc, s| {
+			let locked_amount = s.locked_amount(now);
+			if locked_amount.is_zero() {
+				return acc;
+			}
+
+			let s_end = s.end().expect("ensured not overflow while adding; qed");
+			if let Some((amount, until)) = acc {
+				Some((amount + locked_amount, until.max(s_end)))
+			} else {
+				Some((locked_amount, s_end))
+			}
+		})
+	}
+
+	fn do_add_vesting_schedule(
+		from: &T::AccountId,
+		to: &T::AccountId,
+		schedule: VestingScheduleOf<T>,
+	) -> DispatchResult {
+		// FIXME: remove `do_claim` workaround after issue #4655 fixed
+		// https://github.com/paritytech/substrate/issues/4655
+		let _ = Self::do_claim(to);
```

Comment:
if do_claim is required to call ensure_lockable, should move it inside ensure_lockable